### ISSUE TITLE
Different results with ESP32 and I2C LCD2004 in browser, VSCode and real experiment 

### ISSUE THREAD

**Issue by labinho** [wokwi/wokwi-features#689]
I use a **ESP32 Dev Kit C V4** Board together with a **LCD2004 with I2C** module and a led. I build the firmware with **PlatformIO IDE** in **VSCode** The objective is to print a simple message "hello" over the serial monitor and LCD. A blinking led is just to indicate the program is running and not stuck in the setup.
### The Problem
I get different results in three environments:
- The [browser simulation](https://wokwi.com/projects/381733767944380417) blinks the led shows serial output and LCD text
- The VSCode simulation blinks the led shows **no** serial output and **no** LCD text _<- This is the problem_
- The real experiment blinks the led and LCD text (serial output not captured)

### Reproduce
1. To reproduce test the web browser version first
2. In VSCode with PlatformIO IDE create a new project with the board **Espressif ESP32 Dev Module** and framework **Arduino**
3. Modify the **platform.ini**:
```
[env:az-delivery-devkit-v4]
platform = espressif32
board = az-delivery-devkit-v4
framework = arduino
lib_deps = marcoschwartz/LiquidCrystal_I2C@^1.1.4
```
4. copy the source code from the browser version to the **main.cpp**
5. create a **diagram.json** according to the browser version.
6. create a **wokwi.toml** that contains:
```
[wokwi]
version = 1
firmware = '.pio/build/az-delivery-devkit-v4/firmware.bin'
elf = '.pio/build/az-delivery-devkit-v4/firmware.elf'
```
7. Compile and build
8. Start the simulation

I expect to see serial output and lcd display. There is no serial output "hello" and the lcd stays turned off.

### Environment
 - OS: Mac OS Sonoma 14.1.2 (23B92)
 - Browser: Safari v17.1.2 (19616.2.9.11.12)
 - VSCode Version: 1.84.2 (Universal)
 - PlatformIO IDE v3.3.1
 - Wokwi Simulator v2.1.0

### Remark
I also experience some license problems. If I try to edit the diagram.json file VSCode prompts the message "You need to upgrade to a paid plan to edit the diagram", although I have an active subscription. Maybe this has something to do with disabled features?

**Comment by urish** [comment]
Thanks for the detailed report!

So browser behaves like the real experiment, and the only difference is in VS Code, right?

Re license - can you press "F1" and "Wokwi: Request a New License" and see if the new license helps?

**Comment by labinho** [comment]
Correct. The browser behaves as expected. The VSCode version differs from it. After requesting a new license I am able to modify the diagram.json. The problem with serial monitor and LCD still remains.
![Bildschirmfoto 2023-12-06 um 14 36 34](https://github.com/wokwi/wokwi-features/assets/28773190/a29963dc-4c6d-4d05-b751-1c89eb43332d)

**Comment by urish** [comment]
Can you please attach .pio/build/esp32dev/firmware.bin and .pio/build/esp32dev/firmware.elf  ?

**Comment by labinho** [comment]
[firmware.zip](https://github.com/wokwi/wokwi-features/files/13583569/firmware.zip)

**Comment by urish** [comment]
Thanks! Unfortunately, I'm not able to reproduce this with neither the project I created locally nor the binary files you attached. 

It seems to work fine in both cases:

![image](https://github.com/wokwi/wokwi-features/assets/892318/1cef791b-64a8-4c24-a185-45936cb178fb)

What's your VS Code version?

**Comment by urish** [comment]
Also, can you try with the following wokwi.toml and check if it makes any difference?

```toml
[wokwi]
version = 1
firmware = '.pio/build/esp32dev/firmware.elf'
elf = '.pio/build/esp32dev/firmware.elf'
```

**Comment by labinho** [comment]
My VSCode version is
Version: 1.84.2 (Universal)
Commit: 1a5daa3a0231a0fbba4f14db7ec463cf99d7768e
Datum: 2023-11-09T10:52:33.687Z
Electron: 25.9.2
ElectronBuildId: 24603566
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
Betriebssystem: Darwin arm64 23.1.0

**Comment by labinho** [comment]
> Also, can you try with the following wokwi.toml and check if it makes any difference?
> 
> ```toml
> [wokwi]
> version = 1
> firmware = '.pio/build/esp32dev/firmware.elf'
> elf = '.pio/build/esp32dev/firmware.elf'
> ```

I tried both paths, the one to the esp32dev and the one to the az-delivery-devkit-v4

**Comment by urish** [comment]
The difference here is not the path - rather that the "firmware" points to the elf file instead of bin

**Comment by labinho** [comment]
Sorry, I overlooked this detail. Changing the link from .elf to .bin doesn't alter the behavior.

**Comment by urish** [comment]
Do other ESP32 projects work? 

You can try this one, it's precompiled and known to work:

https://github.com/wokwi/esp-idf-hello-world

**Comment by labinho** [comment]
I figure out how to use the wokwi-cli and come back to you

**Comment by urish** [comment]
That's a VS Code project (check out the part where it says "Running with VS Code")

**Comment by labinho** [comment]
![Bildschirmfoto 2023-12-06 um 15 23 37](https://github.com/wokwi/wokwi-features/assets/28773190/e4c14c4c-d53c-4e9e-959f-cb850365a979)


with the correct token

**Comment by urish** [comment]
Is the LED something you added? The original project didn't have an LED in the diagram

**Comment by labinho** [comment]
I noticed it, too. Somehow VSCode/PlatformIO is strange regarding referencing the correct diagram.json. When I prompt ">wokwi: start simulator" it seems to open another diagram. I have to open the according diagram.json and start the simulation. Using the VSCode command palette is error prone. After starting the [esp-idf-hello-world](https://github.com/wokwi/esp-idf-hello-world) from the diagram, I get the following result
![Bildschirmfoto 2023-12-06 um 15 38 20](https://github.com/wokwi/wokwi-features/assets/28773190/7212e6ec-745c-46f2-9c0a-610dc938ecd8)

**Comment by urish** [comment]
Yes, there are two suspicious things about the screen shot:

1. There are two copies of the Serial terminal. 
2. It doesn't get past the boot sequence.

I'm not yet sure what's happening, but obviously the simulation session is not started correctly. I'll look into how we can further investigate it.

**Comment by labinho** [comment]
Thank you so far for your effort to solve this issue! I leave it as it is for now. I used serial monitor and LCD before in other test projects and can confirm it was working. I try to use other boards and simulation setups to also further investigate the problem if the issue is dependent on the board or the toolchain.

**Comment by labinho** [comment]
I tried the simulator with different configurations. The Arduino nano together with LCD2004/I2C is working fine.

I set up various types of ESP32 with a minimal working program just blinking a led. According to the documentation for [ESP32 Simulation](https://docs.wokwi.com/de-DE/guides/esp32) I additionally built the firmware for platforms S2 and S3 to ensure supported simulation models.
```
esp32dev               SUCCESS   00:00:04.385
esp32-s2-saola-1       SUCCESS   00:00:02.985
esp32s3box             SUCCESS   00:00:03.712
az-delivery-devkit-v4  SUCCESS   00:00:03.781
```
The simulation still does not exceed the booting phase and reboots infinitely. The Terminal appears twice but only after the second simulation run.

 ![Bildschirmfoto 2023-12-06 um 19 07 10](https://github.com/wokwi/wokwi-features/assets/28773190/e6137d64-9192-47b8-be96-c1bcb53d328c)

![Bildschirmfoto 2023-12-06 um 19 07 21](https://github.com/wokwi/wokwi-features/assets/28773190/9f1ca3ee-6b6f-4c21-b5ba-390d1ec89afb)

I have no idea what could be causing the error.

**Comment by urish** [comment]
esp32-s2-saola-1 / esp32s3box binaries won't work with the standard ESP32 board, as the chips are different (you'd need to switch to the board-esp32-s2-devkitm-1 / board-esp32-s3-devkitc-1 to go with them).

Also, good spot re the double terminal - I reproduced it, should be fixed soon.

**Comment by labinho** [comment]
I managed to solve the problem somehow.  I repeatedly did a full clean of the project and switched between environments `esp32dev` and `az-delivery-devkit-v4`. After some attempts both of them work. I don't understand what caused the problem in the first place although I followed the same steps as before. I get the feeling it is a plain user error by mishandling the IDE. The workflow of PlatformIO is distracting and the resulting errors are hard to spot.

Maybe it is worth to log reboots and crashes to find out at which stage the simulation breaks and which model component is affected.

Thank you for your support and your good work!  I very much appreciate it.
